cas-compute/src/symbolic/derivative.rs

```diff
@@ -0,0 +1,268 @@
+use rug::Integer;
+
+use crate::primitive::{float, int};
+
+use super::expr::Primary;
+use super::Expr;
+
+// Something that is not trivially zero - intended to clean up ASTs and not mathetmatically
+// rigorous
+fn non_zero(e: &Expr) -> bool {
+    match e {
+        Expr::Primary(Primary::Float(f)) => {
+            !f.is_zero()
+        },
+        Expr::Primary(Primary::Integer(i)) => {
+            !i.is_zero()
+        },
+        Expr::Add(sum) => {
+            sum.iter().any(non_zero)
+        },
+        Expr::Mul(mul) => {
+            mul.iter().all(non_zero)
+        }
+        _ => true
+    }
+}
+
+// Something that is not trivially one - intended to clean up ASTs and not mathetmatically
+// rigorous
+fn non_unity(e: &Expr) -> bool {
+    match e {
+        Expr::Primary(Primary::Float(f)) => !f.eq(&float(1)),
+        Expr::Primary(Primary::Integer(i)) => !i.eq(&int(1)),
+        Expr::Mul(exprs) => {
+            exprs.iter().any(non_unity)
+        },
+        Expr::Exp(expr, expr1) => {
+            non_zero(expr1) && non_unity(expr)
+        },
+        _ => true
+    }
+}
+
+#[derive(Default)]
+struct MultBuilder(Vec<Expr>);
+
+impl From<MultBuilder> for Expr {
+    fn from(value: MultBuilder) -> Self {
+        if value.0.len() > 1 {
+            Expr::Mul(value.0)
+        } else if value.0.len() == 1 {
+            value.0[0].clone()
+        } else {
+            Expr::Primary(Primary::Integer(int(1)))
+        }
+    }
+}
+
+impl MultBuilder {
+    fn mult(&mut self, e: Expr) {
+        if !non_zero(&e) || self.0.get(0).is_some_and(|e| !non_zero(e)) {
```

Comment:
I find it a bit confusing to mentally process multiple negations in my head. For example, it takes a bit of overhead to go from `!non_zero(&e)` to "not non zero" to "is zero." You may want to consider "reversing" the logic of the helper functions, e.g., by changing `non_zero` to `is_zero` and changing `non_unity` to `is_one`.

Reply:
Good point. Did feel silly to type out the double negative